src/javascript/crypto/e2e/coname/client.js

```diff
@@ -0,0 +1,338 @@
+                pubKey: keys[id].ed25519,
+                privKey: []
+              });
+        }
+      }
+
+      // aggregate (nested) quorums and flatten them in an array
+      realm.verification_policy.quorum_list = e2e.coname.flattenRealmQuorums_(
+          realm.verification_policy.quorum);
+
+      e2e.coname.realmConfig_[domain] = ret = realm;
+
+      return true;
+    }
+  });
+
+  return ret;
+};
+
+
+/**
+ * Get the realm constants based on the domain of email address
+ * @param {string} user The username
+ * @return {?e2e.coname.RealmConfig} The RealmConfig
+ */
+e2e.coname.getRealmByEmail = function(user) {
+  var i = user.indexOf('@');
+  if (i === -1) {
+    throw new e2e.error.InvalidArgumentsError(
+        'GetRealm: user must be of the form .*@.* (got ' + user + ')');
+  }
+  return e2e.coname.getRealmByDomain_(user.slice(i + 1));
+};
+
+
+
+/**
+ * Constructor for the coname client.
+ * @constructor
+ */
+e2e.coname.Client = function() {
+  /**
+   * @type {?Object<string,*>}
+   */
+  this.proto = null;
+
+  /**
+   * @private {boolean}
+   */
+  this.initialized_ = false;
+};
+
+
+/**
+ * Initializes the external protobuf dependency.
+ * @param {function()} callback The function to call when protobuf is loaded.
+ * @param {function(Error)} errback The error callback.
+ */
+e2e.coname.Client.prototype.initialize = function(callback, errback) {
+  var pbURL = chrome.runtime.getURL('protobuf-light.alldeps.js');
+  var clientProtoURL = chrome.runtime.getURL('coname-client.proto.json');
+
+  this.initialized_ = true;
+
+  if (window.dcodeIO && window.dcodeIO.ByteBuffer &&
+      window.dcodeIO.Long && window.dcodeIO.ProtoBuf) {
+    callback.call(this);
+    return;
+  }
+
+  // XXX: jsloader has spurious timeout errors, so set it to 0 for no timeout.
+  goog.net.jsloader.load(pbURL, {timeout: 0}).addCallback(function() {
+    var d = window.dcodeIO, self = this;
+    if (d && d.Long && d.ByteBuffer && d.ProtoBuf) {
+      // Success
+      d.ProtoBuf.loadJsonFile(clientProtoURL, function(err, builder) {
+        if (!err) {
+          self.proto = builder.build('proto');
+          callback.call(self);
+        }
+      });
+    } else {
+      return new Error('Missing protobuf!');
+    }
+  }, this).addErrback(errback, this);
+};
+
+
+/**
+ * @private
+ * Parse, decode, and transform a lookup response
+ * @param {string} jsonString The lookup message to decode
+ * @return {e2e.coname.LookupResponse} The lookup response
+ */
+e2e.coname.Client.prototype.decodeLookupMessage_ = function(jsonString) {
+  var proto = this.proto;
+  var lookupProof = JSON.parse(jsonString);
+  var b64decode = goog.crypt.base64.decodeStringToByteArray;
+
+  // a lot of convertions before they can be verified
+  // TODO: except the use of encodeAB(), may be better to move them to server
+  lookupProof.index = b64decode(lookupProof.index);
+  lookupProof.index_proof = b64decode(lookupProof.index_proof);
```

Comment:
similarly, in case it throws an error, it will be caught at https://github.com/yahoo/end-to-end/pull/66/files#diff-f5016d31de03a18848b49adb2555d420R308